diff --git a/src/api/audio_codecs/audio_format.h b/src/api/audio_codecs/audio_format.h
index edccc17..023ddb6 100644
--- a/src/api/audio_codecs/audio_format.h
+++ b/src/api/audio_codecs/audio_format.h
@@ -25,7 +25,7 @@ namespace webrtc {
 
 // SDP specification for a single audio codec.
 struct RTC_EXPORT SdpAudioFormat {
-  using Parameters [[deprecated(("Use webrtc::CodecParameterMap"))]] =
+  using Parameters [[deprecated("Use webrtc::CodecParameterMap")]] =
       std::map<std::string, std::string>;
 
   SdpAudioFormat(const SdpAudioFormat&);
diff --git a/src/api/video_codecs/sdp_video_format.h b/src/api/video_codecs/sdp_video_format.h
index af9537b..9ee3ed1 100644
--- a/src/api/video_codecs/sdp_video_format.h
+++ b/src/api/video_codecs/sdp_video_format.h
@@ -26,7 +26,7 @@ namespace webrtc {
 // SDP specification for a single video codec.
 // NOTE: This class is still under development and may change without notice.
 struct RTC_EXPORT SdpVideoFormat {
-  using Parameters [[deprecated(("Use webrtc::CodecParameterMap"))]] =
+  using Parameters [[deprecated("Use webrtc::CodecParameterMap")]] =
       std::map<std::string, std::string>;
 
   explicit SdpVideoFormat(const std::string& name);
